File: R/specs.R
source("user_fit.R")

# Generate specifications.
#
#  Arguments:
#      data -- The meta-analytic dataset.
#      which_lists -- The which-factors.
#      how_lists -- The how-factors.
#      colmap -- The column-map from the configuration.
#      k_min -- The minimum number of effects to include a specification.
#      level -- The meta-analysis level (2 or 3).
#      save_path -- The path to where the specification data should be
#                   stored.
#
#  Returns:
#      The specification data as a data frame.
#
generateSpecs <- function(data, which_lists, how_lists, colmap, k_min,
                          level, save_path) {
  # Generate all combinations of which/how factors and get their
  # total number (the amount of specifications)
  group_factors <- c(which_lists, how_lists)
  specs <- expand.grid(group_factors)
  n_specs <- nrow(specs)

  # Add to-be-filled columns with NA values
  tbf_keys <- list(
    "mean", "lb", "ub", "p", "k", "set"
  )
  na_cols <- list()
  for (key in tbf_keys) {
    na_cols[[key]] <- rep(NA, n_specs)
  }

  # Create specification data frame
  specs <- data.frame(specs, na_cols)

  # Get relevant keys from colmap
  key_c_id <- colmap$key_c_id
  key_e_id <- colmap$key_e_id

  # Iterate over all specifications and filter the data according to
  # the which-factors, then perform meta-analytic computations according to
  # how-factors.
  for (i in 1:n_specs) {
    # Copy data into temporary variable
    temp_data <- data

    # Select specification
    spec <- specs[i, ]

    # Iterate over all which-factors and filter data that fits
    for (key in names(which_lists)) {
      # Get specification value for which-factor
      spec_val <- spec[, key]

      # Do not filter data if value is an all-value
      if (startsWith(as.character(spec_val), "all_")) {
        next
      }

      # Filter
      temp_data <- temp_data[temp_data[[key]] == spec_val, ]
    }

    # Disregard data subsets with less than k_min studies
    if (nrow(temp_data) < k_min) {
      next
    }

    # Get sets of cluster- and effect IDs that fulfill the specification
    c_ids <- sort(unique(temp_data[[key_c_id]]))
    e_ids <- sort(temp_data[[key_e_id]])
    # Store IDs as comma-joined lists in specification data
    study_set <- paste(c_ids, collapse = ",")
    es_set <- paste(e_ids, collapse = ",")
    specs$set[i] <- study_set
    specs$set_es[i] <- es_set
    # Store the amount of clusters (kc) and effect sizes (k) that
    # contribute to the specification
    specs$k[i] <- nrow(temp_data)
    specs$kc[i] <- length(c_ids)

    # Fit meta-analytic model with filtered data according to how-factors
    how_values <- list()
    for (k in names(how_lists)) {
      how_values[[k]] <- spec[[k]]
    }
    if (level == 2) {
      res <- fitModelLvl2(how_values, temp_data, colmap)
    } else if (level == 3) {
      res <- fitModelLvl3(how_values, temp_data, colmap)
    }

    # Save model results in specifications data frame
    specs[i, names(res)] <- unlist(res)
  }

  # Remove all specifications which were not used
  specs_full <- specs[complete.cases(specs), ]

  # Only keep unique study/sample subsets
  cols <- c("mean", "set", names(how_lists))
  specs_full <- specs_full[!duplicated(specs_full[, cols]), ]

  # Add binary indicator if all studies are included in the subset
  c_ids <- sort(unique(data[[key_c_id]]))
  all_studies_list <- paste(c_ids, collapse = ",")
  full_set <- (specs_full$set == all_studies_list)
  specs_full$full_set <- as.numeric(full_set)

  # Rank specifications according to mean in ascending order
  specs_full$rank <- rank(specs_full$mean, ties.method = "random")
  # Store CI interval width in dedicated column
  specs_full$ci <- specs_full$ub - specs_full$lb

  # !important Sort specifications by rank
  specs_full <- specs_full[order(specs_full$rank), ]

  # Save specs to CSV file
  write.table(
    specs_full,
    file = save_path,
    row.names = FALSE,
    dec = ".",
    sep = ","
  )

  return(specs_full)
}

File: Python/specs.py
from itertools import product
import numpy as np
import pandas as pd
from tqdm import tqdm

from user_fit import fit_model_lvl_2, fit_model_lvl_3


def generate_specs(data, which_lists, how_lists, colmap, k_min, level,
                   save_path):
    """Generate specifications.

    Arguments:
        data -- The meta-analytic dataset.
        which_lists -- The which-factors.
        how_lists -- The how-factors.
        colmap -- The column-map from the configuration.
        k_min -- The minimum number of effects to include a specification.
        level -- The meta-analysis level (2 or 3).
        save_path -- The path to where the specification data should be
                     stored.

    Returns:
        The specification data as a pandas DataFrame.
    """
    # Generate all combinations of which/how factors and get their
    # total number (the amount of specifications)
    group_factors = dict(which_lists, **how_lists)
    cart_prod = product(*[values for values in group_factors.values()])
    specs = pd.DataFrame(cart_prod, columns=list(group_factors.keys()))
    n_specs = len(specs)

    # Add to-be-filled columns with None values
    tbf_keys = ["mean", "lb", "ub", "p", "k", "set"]
    for key in tbf_keys:
        specs[key] = [None] * n_specs

    # Get relevant keys from colmap
    key_c_id = colmap["key_c_id"]
    key_e_id = colmap["key_e_id"]

    # Iterate over all specifications and filter the data according to
    # the which-factors, then perform meta-analytic computations according to
    # how-factors.
    for i in tqdm(range(n_specs)):
        # Copy data into temporary variable
        temp_data = data

        # Select specification
        spec = specs.iloc[i]

        # Iterate over all which-factors and filter data that fits
        for key in which_lists.keys():
            # Get specification value for which-factor
            spec_val = spec[key]

            # Do not filter data if value is an all-value
            if spec_val.startswith("all_"):
                continue

            # Filter
            temp_data = temp_data[temp_data[key] == spec_val]

        # Disregard data subsets with less than k_min studies
        if len(temp_data) < k_min:
            continue

        # Get sets of cluster- and effect IDs that fulfill the specification
        c_ids = sorted(temp_data[key_c_id].unique())
        e_ids = sorted(temp_data[key_e_id])
        # Store IDs as comma-joined lists in specification data
        study_set = (",").join([str(c_id) for c_id in c_ids])
        es_set = (",").join([str(e_id) for e_id in e_ids])
        specs.at[i, "set"] = study_set
        specs.at[i, "set_es"] = es_set
        # Store the amount of clusters (kc) and effect sizes (k) that
        # contribute to the specification
        specs.at[i, "k"] = len(temp_data)
        specs.at[i, "kc"] = len(c_ids)

        # Fit meta-analytic model with filtered data according to how-factors
        how_values = {k: spec[k] for k in how_lists.keys()}
        if level == 2:
            res = fit_model_lvl_2(
                how_values,
                temp_data,
                colmap
            )
        elif level == 3:
            res = fit_model_lvl_3(
                how_values,
                temp_data,
                colmap
            )

        # Save model results in specifications data frame
        specs.at[i, res.keys()] = res.values()

    # Remove all specifications which were not used
    specs_full = specs.dropna(axis=0, how="any")

    # Only keep unique study/sample subsets # TODO
    specs_full = specs_full.drop_duplicates(
        subset=["mean", "set", *how_lists.keys()]
    )

    # Add binary indicator if all studies are included in the subset
    c_ids = sorted(data[key_c_id].unique())
    all_studies_list = (",").join([str(c_id) for c_id in c_ids])
    full_set = (specs_full["set"] == all_studies_list)
    specs_full["full_set"] = full_set.astype("int64")

    specs_full.reset_index(drop=True, inplace=True)

    # Rank specifications according to mean in ascending order
    specs_full["rank"] = np.argsort(np.argsort(specs_full["mean"])) + 1
    # Store CI interval width in dedicated column
    specs_full["ci"] = specs_full["ub"] - specs_full["lb"]

    # !important Sort specifications by rank
    specs_full.sort_values(by="rank", inplace=True)

    # Ensure correct datatypes in pandas DataFrame
    for key in ["mean", "lb", "ub", "ci", "p"]:
        specs_full[key] = specs_full[key].astype("float64")
    for key in ["k", "kc"]:
        specs_full[key] = specs_full[key].astype("int64")

    # Save specs to CSV file
    specs_full.to_csv(save_path, index=False)

    return specs_full
